readme.md
# Quickhost

Server and upload-client to quickly host a single static html page - useful for [browserstack](https://browserstack.com) test.

## Motivation

To test JavaScript code running in a browser, browserStack offers a good test platform but the selenium tests require the html
files to be available in the internet. The problem is if you have files locally: how can you make them accessible by browserstack?

This is a simple http hosting server that - using a secret - allows to upload any given html file and hosts it for a limited amount
of time. It also has a limit on amount of files to be hosted which means it cleans up after itself.

## Start a server

Prerequisites: Node running!

```sh
$ env \
    QUICKHOST_SECRET=$(node -p "crypto.randomBytes(8).toString('hex')") \
    QUICKHOST_MAX=10 # (optional) Max number of html files to be hosted - default 10 \
    QUICKHOST_MAX_SIZE=3145728 # (optional) Max number of bytes of a file to be hosted - default 3 Megabytes \
    PORT=1234 # (optional) Port where the server will connect to - default 1234 \
    HOST=localhost # (optional) Host where the server will connect to - default 0 \
    npx quickhost
```

**Important note:** You should run nginx and letsencrypt to have turn the [server into an https server](https://www.codementor.io/@marcoscasagrande/installing-express-nginx-app-on-ubuntu-18-04-with-ssl-using-certbot-pdt44g5gs).

## Upload

Once the server is running you can upload and download files from that server using a simple API:

1. `npm install @leichtgewicht/quickhost`
2. 
    ```javascript
    const upload = require('quickhost')
    const download = require('node-fetch')
    const server = 'https://quickhost.my-host.com'

    const key = await upload({
      server, // Server url where the quickhost server is running
      secret: 'secret-from-server', // Secret used when starting the quickhost server
      data: '<html><b>Hello World</b></html>', // Data to be uploaded to the quickhost server
      timeout: 5000 // Timeout for the upload
    })

    const data = await fetch(`${server}/${key}`) // Now we can download the server
    ```

## License

[MIT](./LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.1.0](https://github.com/martinheidegger/quickhost/compare/v1.0.0...v1.1.0) (2020-06-02)


### Features

* Added content-header support for hosted files. ([b8894cf](https://github.com/martinheidegger/quickhost/commit/b8894cfc6361ab880b6167d4001147d4db81135c))
* JSON output for errors ([ada824d](https://github.com/martinheidegger/quickhost/commit/ada824da87807b50e2565c6cbca83e1d68428d4d))


test.js
const { test } = require('tap')
const fetch = require('node-fetch')
const http = require('http')
const _createServer = require('./server.js')
const upload = require('./upload.js')
const { AbortController } = require('abortcontroller-polyfill/dist/cjs-ponyfill')
const series = require('p-series')

async function getDataAndType (server, key, opts) {
  const res = await fetch(`${server}/${key}`, opts)
  if (res.status !== 200) {
    throw Object.assign(new Error(`[EHTTPSTATUS] [statusCode=${res.status}] ${await res.text()}`), { statusCode: res.status, code: 'EHTTPSTATUS' })
  }
  return { data: await res.text(), type: res.headers.get('content-type') }
}

async function getData (server, key, opts) {
  return (await getDataAndType(server, key, opts)).data
}

async function createServer (opts) {
  const control = new AbortController()
  const server = await _createServer({
    ...opts,
    abortSignal: control.signal
  })
  return {
    ...server,
    control,
    server: `http://localhost:${server.address.port}`,
    close () {
      const p = server.closePromise.catch(() => {})
      control.abort()
      return p
    }
  }
}

test('basic functionality', async t => {
  const { server, close } = await createServer({
    max: 1,
    secret: 'abcd'
  })
  try {
    const key = await upload({ server, secret: 'abcd', data: 'hello world', timeout: 5000 })
    t.match(key, /^[0-9a-f]{12}$/)
    t.equals(await getData(server, key), 'hello world')
  } finally {
    await close()
  }
})

test('requesting via post should throw error', async t => {
  const { server, close } = await createServer({
    max: 1,
    secret: 'abcd'
  })
  try {
    const key = await upload({ server, secret: 'abcd', data: 'hello world', timeout: 5000 })
    t.match(key, /^[0-9a-f]{12}$/)
    await rejects(t, getData(server, key, { method: 'POST' }), err => t.equals(err.statusCode, 404))
  } finally {
    await close()
  }
})

test('non existing file', async t => {
  const { server, close } = await createServer({
    max: 1,
    secret: 'abcd'
  })
  try {
    await rejects(t, getData(server, 'test'), err => t.equals(err.statusCode, 404))
  } finally {
    await close()
  }
})

test('uploading too much', async t => {
  const { server, close } = await createServer({
    max: 1,
    maxSize: 1,
    secret: 'abcd'
  })
  try {
    await upload({ server, secret: 'abcd', data: 'hello world', timeout: 5000 })
    t.fail('error')
  } catch (err) {
    t.equals(err.code, 'EHTTPSTATUS')
    t.equals(err.statusCode, 413)
    t.equals(err.message, JSON.stringify({ code: 413, message: 'payload too large', size: 11, maxSize: 1 }, null, 2))
  } finally {
    await close()
  }
})

test('max amount is respected', async t => {
  const { server, close } = await createServer({
    max: 2,
    secret: 'abcd'
  })
  try {
    const keys = await series([
      () => upload({ server, secret: 'abcd', data: 'test-a', timeout: 5000 }),
      () => upload({ server, secret: 'abcd', data: 'test-b', timeout: 5000 }),
      () => upload({ server, secret: 'abcd', data: 'test-c', timeout: 5000 })
    ])
    t.equals(await getData(server, keys[1]), 'test-b')
    t.equals(await getData(server, keys[2]), 'test-c')
    try {
      const data = await getData(server, keys[0])
      console.log({ data })
      t.fail('error')
    } catch (err) {
      t.equals(err.code, 'EHTTPSTATUS')
      t.equals(err.statusCode, 404)
    }
  } finally {
    await close()
  }
})

test('upload client timeout', async t => {
  const { server, close } = await createServer({
    max: 2,
    secret: 'abcd'
  })
  try {
    await upload({ server, secret: 'abcd', timeout: 1, data: 'hello world' })
    t.fail('error')
  } catch (err) {
    t.equals(err.code, 'ETIMEOUT', err.message)
    t.equals(err.timeout, 1)
  } finally {
    await close()
  }
})

test('broken server config', async t => {
  await rejects(t, createServer({}), 'EARG')
  await rejects(t, createServer({ secret: 'abcd' }), 'EARG')
  await rejects(t, createServer({ secret: 'abcd', max: 0 }), 'EARG')
  await rejects(t, createServer({ secret: 'abcd', max: 1, maxSize: 0 }), 'EARG')
})

test('upload server timeout', async t => {
  const { server, close } = await createServer({
    max: 2,
    secret: 'abcd',
    timeout: 1
  })
  try {
    await (new Promise((resolve) => {
      const req = http.request(`${server}/abcd`, { method: 'POST' }, res => {
        t.equals(res.statusCode, 408)
        const result = []
        res.on('data', data => result.push(data))
        res.on('error', err => {
          t.fail('Unexpected error' + err)
          req.close()
          resolve()
        })
        res.on('close', () => {
          const message = Buffer.concat(result).toString()
          t.equals(message, JSON.stringify({ code: 408, message: 'timeout', timeout: 1 }, null, 2))
          resolve()
        })
      })
      req.on('error', () => {})
      req.write('msg')
    }))
  } finally {
    await close()
  }
})

test('custom content type', async t => {
  const { server, close } = await createServer({
    max: 1,
    secret: 'abcd'
  })
  try {
    const key = await upload({ server, secret: 'abcd', data: JSON.stringify('hello world'), contentType: 'application/json', timeout: 5000 })
    t.match(key, /^[0-9a-f]{12}$/)
    const { data, type } = await getDataAndType(server, key)
    t.equals(type, 'application/json')
    t.equals(data, '"hello world"')
  } finally {
    await close()
  }
})

async function rejects (t, promise, test) {
  let thrown = false
  try {
    await promise
  } catch (err) {
    thrown = true
    if (typeof test === 'function') {
      test(err, t)
    } else {
      const code = typeof test === 'string' ? test : test.code
      t.equals(err.code, code, err.message)
    }
  }
  if (!thrown) {
    t.fail('Expected rejection')
  }
}


upload.js
const https = require('https')
const http = require('http')

class TimeoutError extends Error {
  constructor (timeout) {
    super(`[ETIMEOUT] Exceeded timeout of ${timeout}`)
    this.timeout = timeout
    this.code = 'ETIMEOUT'
  }
}
module.exports = ({ server, secret, data, contentType, timeout }) => {
  const system = /^https/.test(server) ? https : http
  return new Promise((resolve, reject) => {
    const result = []
    const onData = buffer => result.push(buffer)
    const onEnd = error => {
      clearTimeout(timer)
      req.off('error', onEnd)
      if (error) {
        reject(error)
      } else {
        resolve(Buffer.concat(result).toString())
      }
    }
    const req = system.request(
      `${server}/${secret}`,
      {
        method: 'POST',
        headers: {
          'store-content-type': contentType || 'text/html'
        }
      },
      res => {
        clearTimeout(timer)
        timer = setTimeout(() => onResEnd(new TimeoutError(timeout)), timeout)
        const onResEnd = error => {
          res.off('error', onResEnd)
          res.off('end', onResEnd)
          res.off('data', onData)
          if (res.statusCode !== 200) {
            const message = Buffer.concat(result).toString()
            onEnd(Object.assign(new Error(`[EHTTPSTATUS] ${res.statusCode}: ${message}`), { statusCode: res.statusCode, code: 'EHTTPSTATUS', message }))
          } else {
            onEnd(error)
          }
        }
        res.on('error', onResEnd)
        res.on('end', onResEnd)
        res.on('data', onData)
      }
    )
    let timer = setTimeout(() => onEnd(new TimeoutError(timeout)), timeout)
    req.on('error', onEnd)
    req.write(data)
    req.end()
  })
}


server.js
const { createServer } = require('http')
const { randomBytes } = require('crypto')
const createLRU = require('lru')
const JSON_TYPE = { 'Content-Type': 'application/json' }
const TEXT_TYPE = { 'Content-Type': 'text/text' }

function res404 (res) {
  res.writeHead(404).end('404 - not found', 'utf-8')
}

module.exports = ({ abortSignal, host, port, max, maxAge, maxSize, timeout, secret }) => new Promise((resolve, reject) => {
  if (typeof secret !== 'string' || secret.length === 0) {
    return reject(Object.assign(new Error(`[EARG] secret is wrong, should be string of length > 0 but is ${secret}`), { code: 'EARG' }))
  }
  if (isNaN(max) || max === null || max < 1) {
    return reject(Object.assign(new Error(`[EARG] Max amount of items needs to be a number, but is ${max}`), { code: 'EARG' }))
  }
  if (isNaN(maxSize) || max === null) {
    maxSize = 1024 * 1024 * 3 // 3 Megabytes
  } else if (maxSize >= 1) {
    maxSize |= 0
  } else {
    return reject(Object.assign(new Error(`[EARG] Max size needs to be a positive integer, but is ${maxSize}`), { code: 'EARG' }))
  }
  if (isNaN(timeout) || timeout === null) {
    timeout = 5000
  }
  const lru = createLRU({ max, maxAge })
  const headerMap = new WeakMap()
  const processUpload = (req, res) => {
    const buffer = []
    let size = 0
    let timer
    const reTimeout = () => {
      clearTimeout(timer)
      timer = setTimeout(() => {
        finish()
        res.writeHead(408, JSON_TYPE).end(JSON.stringify({ code: 408, message: 'timeout', timeout }, null, 2))
      }, timeout)
    }
    const onData = data => {
      reTimeout()
      size += data.byteLength
      if (size > maxSize) {
        finish()
        res.writeHead(413, JSON_TYPE).end(JSON.stringify({ code: 413, message: 'payload too large', size, maxSize }, null, 2))
        return
      }
      buffer.push(data)
    }
    const onEnd = () => {
      finish()
      const key = randomBytes(6).toString('hex')
      const content = Buffer.concat(buffer)
      lru.set(`/${key}`, content)
      headerMap.set(content, { 'Content-Type': req.headers['store-content-type'] || 'text/html' })
      res.writeHead(200, TEXT_TYPE).end(key)
    }
    const finish = () => {
      clearTimeout(timer)
      req.off('data', onData)
      req.off('end', onEnd)
      req.off('aborted', finish)
      req.off('error', finish)
    }
    req.on('aborted', finish)
    req.on('error', finish)
    req.on('data', onData)
    req.on('end', onEnd)
    reTimeout()
  }
  const isAuthorized = req => {
    return req.method === 'POST' && req.url === `/${secret}`
  }
  const server = createServer((req, res) => {
    if (isAuthorized(req)) {
      processUpload(req, res)
    } else if (req.method !== 'GET') {
      res404(res)
    } else {
      const content = lru.get(req.url)
      if (content === undefined) {
        res404(res)
      } else {
        res.writeHead(200, headerMap.get(content)).end(content)
      }
    }
  })
  const closePromise = new Promise((resolve, reject) => {
    server.once('error', (error) => {
      listener.close()
      reject(error)
    })
    server.once('close', () => {
      reject(Object.assign(new Error('[ERRABORT]'), { code: 'ERRABORT' }))
    })
  })
  server.once('error', reject)
  server.once('close', () => {
    reject(Object.assign(new Error('[ERRABORT]'), { code: 'ERRABORT' }))
  })
  server.once('listening', () => {
    resolve({ closePromise, address: listener.address(), secret })
  })
  const listener = server.listen(port, host)
  if (abortSignal) {
    abortSignal.addEventListener('abort', () => server.close())
  }
})
